### ISSUE TITLE
calculate the perplexity on penntreebank using LSTM keras got infinity

### ISSUE THREAD

**Issue by janenie** [keras-team/keras#2317]
I am very new to KERAS, and I use the dealt dataset from the RNN Toolkit and try to use LSTM to train the language model
I have problem with the calculating the perplexity though. It always get quite large negative log loss, and when using the exp function, it seems to get the infinity, I got stuck here.
Below is my model code, and the github link( https://github.com/janenie/lstm_issu_keras ) is the current problematic code of mine. Can someone help me out?

the test_y data format is word index in sentences per sentence per line, so is the test_x. 

class LSTMLM:
    def **init**(self, input_len, hidden_len, output_len, return_sequences=True):
        self.input_len = input_len
        self.hidden_len = hidden_len
        self.output_len = output_len
        self.seq = return_sequences
        self.model = Sequential()

```
def build(self, maxlen=50, dropout=0.2):
    self.model.add(Embedding(self.input_len, self.hidden_len, input_length=maxlen))
    self.model.add(LSTM(output_dim=self.hidden_len, return_sequences=True))
    #self.model.add(Dropout(dropout))
    self.model.add(TimeDistributedDense(self.output_len))
    self.model.add(Activation('softmax'))
    self.model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

def train(self, train_x, train_y, dev_x, dev_y, batchsize=16, epoch=1):
    hist = self.model.fit(train_x, train_y, batch_size=batchsize, nb_epoch=1, \
                    show_accuracy=True, validation_data=(dev_x, dev_y))
    print hist.history

def saveModel(self, save_file):
    self.model.save_weights(save_file)

def computePPL(self, test_x, test_y, maxLen=100):
    predictions = self.model.predict(test_x)
    test_num = len(test_y)
    maxlen = predictions.shape[1]
    ppl = 0.0
    total_words = 0
    for i in xrange(test_num):
        pred_idx = test_y[i]
        #print np.max(pred_idx)
        sent_len = len(test_y[i])
        padding = maxlen - sent_len
        for j in xrange(sent_len):
            prob_ = predictions[i, j+padding, pred_idx[j]]
            ppl -= np.log(prob_)
        total_words += sent_len
    return np.exp(ppl)/(float(total_words)) 
```

Please make sure that the boxes below are checked before you submit your issue. Thank you!

**Comment by braingineer** [comment]
You can add perplexity as a metric as well:

```python
def perplexity(y_true, y_pred, mask=None):
    if mask is not None:
        y_pred /= K.sum(y_pred, axis=-1, keepdims=True)
        mask = K.permute_dimensions(K.reshape(mask, y_true.shape[:-1]), (0, 1, 'x'))
        truth_mask = K.flatten(y_true*mask).nonzero()[0]  ### How do you do this on tensorflow?
        predictions = K.gather(y_pred.flatten(), truth_mask)
        return K.pow(2, K.mean(-K.log2(predictions)))
    else:
        return K.pow(2, K.mean(-K.log2(y_pred)))
```

though, this doesn't work on tensor flow because I'm only using Theano and haven't figured out how  `nonzero()` works in tensorflow yet. 

In general, though, you average the negative log likelihoods, which forms the empirical entropy (or, mean loss).  This is the quantity used in perplexity.  Additionally, perplexity shouldn't be calculated with e.  It should be calculated as 2 ** L using a base 2 log in the empirical entropy. 

See [Socher's notes here](cs224d.stanford.edu/lecture_notes/LectureNotes4.pdf), [the wikipedia entry](https://en.wikipedia.org/wiki/Perplexity#Perplexity_per_word), and [a classic paper on the topic](http://www.speech.sri.com/projects/srilm/manpages/pdfs/chen-goodman-tr-10-98.pdf) for more information.

**Comment by cheetah90** [comment]
@braingineer Thanks for the code! Unfortunately, the `log2()` is not available in [Keras' backend API] (https://keras.io/backend/). Is there another way to do that? (Or is the `log2()`going to be included in the next version of Keras?)

**Comment by braingineer** [comment]
hey @cheetah90 

rather than futz with things (it's not implemented in tensorflow), you can approximate log2.   log_2(x) = log_e(x)/log_e(2)

so, precompute 1/log_e(2) and just multiple it by log_e(x). 

the following should work (I've used it personally):

```python
oneoverlog2 = 1.442695
result = K.log(x)*oneoverlog2
```

**Comment by cheetah90** [comment]
Hi @braingineer. Thanks! Yeah, I should have thought about that myself :)
Btw, I looked at the Eq8 and Eq9 in [Socher's notes](http://cs224d.stanford.edu/lecture_notes/LectureNotes4.pdf), and actually implemented it differently. 

```
def perplexity(y_true, y_pred):
    time_steps = y_pred.shape[0]
    cross_entropy_over_time = - K.dot(T.flatten(y_true), T.log2(T.flatten(y_pred))) / time_steps
    return pow(2, cross_entropy_over_time)
```
(Of course, my code has to import Theano which is suboptimal in this case. But let me know if there is other way to leverage the `T.flatten` function since it's not in the Keras' backend either)  

Let me know what you think! Thank you!

**Comment by braingineer** [comment]
Hi @cheetah90 

That won't take into account the mask. It's for fixed-length sequences. 

There's a nonzero operation that requires theano anyway in my version. Now that I've played more with Tensorflow, I should update it. I have some deadlines today before I have time to do that, though.  I'll try to remember to comment back later today with a modification

**Comment by cheetah90** [comment]
That's right! It's for the fixed-length and thanks for telling me what the `Mask` means - I was curious about that so didn't implement it. Yeah I will read more about the use of Mask! But anyway, I think according to Socher's note, we will have to dot product the `y_pred` and `y_true` and average that for all vocab in all times.

**Comment by nmoenning** [comment]
Hi @braingineer 

Thanks for sharing your code snippets! I went with your implementation and the little trick for 1/log_e(2). Seems to work fine for me. I wondered how you actually use the `mask` parameter when you give it to `model.compile(..., metrics=[perplexity])`?

**Comment by icoxfog417** [comment]
Accordings to the Socher's notes that is presented by @cheetah90 , could we calculate perplexity by following simple way?

```py
def perplexity(y_true, y_pred):
        cross_entropy = K.categorical_crossentropy(y_pred, y_true)
        perplexity = K.exp(cross_entropy)
        return perplexity
```

**Comment by HotAllure** [comment]
Hi, everyone.

I implemented perplexity according to @icoxfog417 's post, and I got same result - perplexity got inf.
But in my case, I set perplexity as a metrics in model.compile(), then perplexity always got inf on training, and got some reasonable value on validation.
Does anyone solve this problem or implement perplexity in other ways?

**Comment by HotAllure** [comment]
Hi.

I found a simple mistake in my code, it's not related to perplexity discussed here.
After changing my code, perplexity according to @icoxfog417 's post works well.

Just a quick report, and hope that anyone who has same problem will  resolve.

**Comment by MohameddSheriff** [comment]
Ok so I implemented the perplexity according to @icoxfog417 , now i need to evaluate the final perplexity of the model on my test set using model.evaluate(), any help is appreciated.

**Comment by okoub** [comment]
@janenie Do you have an example of how to use your code to create a language model and check it's perplexity?

**Comment by icoxfog417** [comment]
I implemented a language model by Keras (tf.keras) and calculate its perplexity. Please refer following notebook (it uses my preprocessing library [chariot](https://github.com/chakki-works/chariot)).

[language modeling](https://github.com/chakki-works/chariot/blob/master/notebooks/language%20modeling.ipynb)

**Comment by manharwalia11** [comment]
> Accordings to the Socher's notes that is presented by @cheetah90 , could we calculate perplexity by following simple way?
> 
> ```python
> def perplexity(y_true, y_pred):
>     cross_entropy = K.categorical_crossentropy(y_true, y_pred)
>     perplexity = K.pow(2.0, cross_entropy)
>     return perplexity
> ```

But what is y_true,, in text generation we dont have y_true

**Comment by OmniaZayed** [comment]
> Accordings to the Socher's notes that is presented by @cheetah90 , could we calculate perplexity by following simple way?
> 
> ```python
> def perplexity(y_true, y_pred):
>     cross_entropy = K.categorical_crossentropy(y_true, y_pred)
>     perplexity = K.pow(2.0, cross_entropy)
>     return perplexity
> ```

@icoxfog417 what is the shape of y_true and y_pred?
